stm32 LPTIM driver counter value without interrupt

FRASTM opened the pull request:
This PR adds a `counter value` property to the stm32 LPTIM instance, 
so that the LPTIM counts without interruption during this period. 
The value is limited to the range defined by the possible prescaler: any value between [1 .. 64] or [1.. 128] depending on the stm32 serie

For example with 
```
&stm32_lp_tick_source {
	st,counter-value = <15>;
	st,prescaler = <8>;
};
```

k_msleep(TIMEOUT > 15);  will not wake up during 15seconds

Setting a `st,counter-value`  will define a _prescaler_ depending on the LPTIM clock source (LSI or LSE). Consequently the 
prescaler must be equal to the one given in the DTS `<st,prescaler>`. The CONFIG_SYS_CLOCK_TICKS_PER_SEC still needs to be set in function of that calculated prescaler.

msmttchr commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,23 @@ static int sys_clock_driver_init(void)
 
+#if DT_PROP(DT_DRV_INST(0), st_counter_value)
+	/* Calculate the corresponding prescaler */
+	for (lptim_clock_presc = 1; lptim_clock_presc <= 32 ; lptim_clock_presc++) {
+		if (lptim_clock_presc > (uint32_t)DT_PROP(DT_DRV_INST(0), st_counter_value)) {
+			break;
+		}
+	}
```
```suggestion
	lptim_clock_presc = MAX(1, (uint32_t)DT_PROP(DT_DRV_INST(0), st_counter_value));
	lptim_clock_presc = MIN(32, lptim_clock_presc);
```
I am not sure whether this faster and more readable, just a suggestion.

FRASTM replied:
_more readable_ for sure

FRASTM replied:
file:///local/home/frq09495/Pictures/Screenshots/Screenshot%20from%202024-02-19%2010-55-57.png

msmttchr replied:
Ok, I was under the impression that the loop could return a value between 1 and 32 and not only power of 2.

FRASTM committed:
drivers: stm32 lptim driver with a exact LPTIM timeout value

With this change, the LPTIM counter will be able to set
its timeout to the st,timeout value. So that system can
sleep for that period without interruption.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: boards: stm32 power lptim adjusted to 15 seconds

Gives an example where the timeout value is 15seconds,
the prescaler of 8 can let the LPTIM count from 0 to 16 seconds
with LSE

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango reviewed (changes requested):
As discussed, setting the prescaler at runtime is not feasible as today, prescaler value is tied to CONFIG_SYS_CLOCK_TICKS_PER_SEC value.
 
Given this limitation, and in order to comply with existing code and settings around that we already have, the new property should only allow to define a value within the range defined by `st,prescaler` setting and we should trigger an error/assert/warning at runtime if not compatible.

FRASTM commented:
With this change:
 the &stm32_lp_tick_source node can set a `st,counter-value` property to count the exact timeout value.
 the &stm32_lp_tick_source node can omit  `st,prescaler` property  :  
Do not set a default prescaler in the st,stm32-lptim.yaml  : that will allow not setting the st,prescaler  for the &stm32_lp_tick_source node. But the driver will define a default prescaler to 1 (divided by 1)

Do not set a default prescaler in the st,stm32-lptim.yaml  : that will allow not setting the st,prescaler  for the &stm32_lp_tick_source node. But the driver will define a default prescaler to 1 (divided by 1)

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,36 @@ static int sys_clock_driver_init(void)
+	__ASSERT(lptim_clock_presc == (1 << presci),
+		"<st_counter_value> and <st_prescaler> do not match");
+
+	lptim_clock_presc = (1 << presci);
```
line removed

FRASTM commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -20,6 +20,16 @@ include:
       - st,countermode
 
 properties:
+  st,counter-value:
+    type: int
+    description: |
+        Gives the LPTIM an exact counting value for timeout expiration
+        Valid range is [1, 128]
+        If set, this value will define a prescaler and an autoreload counter
+        This must fit the st,prescaler property as defined below.
+        An error is raised if mismatch.
+        The CONFIG_SYS_CLOCK_TICKS_PER_SEC must still be adjusted.
```
changed to
```
  st,counter-value:
    type: int
    description: |
        Gives the LPTIM an exact counting value for timeout expiration.
        Valid range is [1, 128] and should be consistent with st,prescaler
        pre-defined setting.
```

FRASTM commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -20,6 +20,16 @@ include:
+  st,counter-value:
+    type: int
+    description: |
+        Gives the LPTIM an exact counting value for timeout expiration
+        Valid range is [1, 128]
+        If set, this value will define a prescaler and an autoreload counter
+        This must fit the st,prescaler property as defined below.
+        An error is raised if mismatch.
+        The CONFIG_SYS_CLOCK_TICKS_PER_SEC must still be adjusted.
```
changed to 

```
  st,counter-value:
    type: int
    description: |
        Gives the LPTIM an exact counting value for timeout expiration.
        Valid range is [1, 128] and should be consistent with st,prescaler
        pre-defined setting.

```

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,35 @@ static int sys_clock_driver_init(void)
+	 * lptim_time_base is the autoreload counter
+	 */
+	lptim_time_base = 2 * (lptim_clock_freq *
+		(uint32_t)DT_PROP(DT_DRV_INST(0), st_counter_value))
```
```
		(uint32_t)DT_INST_PROP(0, st_counter_value))
```

FRASTM replied:
changed to `DT_INST_PROP(0, st_counter_value)`

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,35 @@ static int sys_clock_driver_init(void)
+	}
+
+	/*
+	 * LPTIM is counting DT_PROP(DT_DRV_INST(0), st_counter_value),
```
```
	 * LPTIM is counting DT_INST_PROP(0, st_counter_value),
```

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,35 @@ static int sys_clock_driver_init(void)
+	}
+
+	/*
+	 * Check if lptim_clock_presc given by DT_PROP(DT_DRV_INST(0), st_prescaler)
+	 * equals the calculated one from the DT_PROP(DT_DRV_INST(0), st_counter_value)
```
```
	 * Check if lptim_clock_presc given by DT_INST_PROP(0, st_prescaler)
	 * equals the calculated one from the DT_INST_PROP(0, st_counter_value)
```

FRASTM replied:
changed to `DT_INST_PROP(0, st_counter_value)`

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -438,6 +467,7 @@ static int sys_clock_driver_init(void)
 		return -EIO;
 	}
 
+#endif /* st_counter_value */
```
Break a line

FRASTM replied:
+ break line

erwango commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -45,3 +45,10 @@ properties:
+  st,counter-value:
+    type: int
+    description: |
+        Gives the LPTIM an exact counting value for timeout expiration.
```
`        Gives the LPTIM an exact counting value (s) for timeout expiration.`

FRASTM replied:
line changed

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,35 @@ static int sys_clock_driver_init(void)
+	/* Calculate the corresponding prescaler in range [1..128] */
+	uint8_t presci;
+
+	for (presci = 0; presci <= 7 ; presci++) {
```
Now that `st_counter_value` could only be with bounds defined by `st_prescaler`, can't we use a faster check ?
For instance  if st_counter_value > (lptim_clock_freq / lptim_clock_presc) * 0xFFFF

erwango commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -45,3 +45,10 @@ properties:
       - 64
       - 128
+
+  st,counter-value:
```
Not sure if the name is clear enough. We can ask around for more explicit names.

FRASTM replied:
what about `st,timeout`

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,35 @@ static int sys_clock_driver_init(void)
+	for (presci = 0; presci <= 7 ; presci++) {
+		/* value is in the interval from 0 to 2<<presc */
+		if ((uint32_t)(2 << presci) >=
+			(uint32_t)DT_PROP(DT_DRV_INST(0), st_counter_value)) {
```
changed to `DT_INST_PROP(0, st_counter_value)`

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -429,6 +429,35 @@ static int sys_clock_driver_init(void)
 	}
 #endif
 
+#if DT_PROP(DT_DRV_INST(0), st_counter_value)
```
changed to `DT_INST_PROP(0, st_counter_value)`

FRASTM commented:
new request : re-opening

FRASTM commented:
rebased on 03959a20f7834ffe50a51df2b2e45ef87d1b610b
No more boards/.conf file required

erwango reviewed (changes requested):
Please fix the clang-format suggestions on new code.
Otherwise LGTM

FRASTM commented:
> Please fix the clang-format suggestions on new code. Otherwise LGTM

done
